File: ServiceCore/Translators/IncommingInvoiceTranslator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using BOCore;
using DALCore;
using DALCore.Models;

namespace ServiceCore.Translators
{
    public class IncommingInvoiceTranslator
    {
        internal static ErrorCode TranslateEntityToBO(IncommingInvoices _entity, IncommingInvoiceBO bo)
        {
            if (_entity == null)
                return ErrorCode.EntityNull;
            if (bo == null)
                return ErrorCode.BoNull;
            bo.Id = _entity.Id;
            bo.ProjectId = _entity.Project.ProjectId;
            if (_entity.Contract is not null)
                bo.ContractID = _entity.ContractId;
            if (_entity.Company is not null)
                bo.CompanyId = _entity.CompanyId;
            bo.IncommingInvoiceDate = _entity.Date;
            if (_entity.ExternalId is not null)
                bo.InvoiceExternalId = _entity.ExternalId;
            bo.InvoicePrice = _entity.Price;
            foreach (var x in _entity.IncommingInvoiceDetail)
            {
                IncommingInvoiceDetailBO bou = new IncommingInvoiceDetailBO();
                var err = IncommingInvoiceDetailTranslator.TranslateEntityToBO(x, bou);
                bo.Details.Add(bou);
            }
            return ErrorCode.Success;
        }
        internal static ErrorCode TranslateBOToEntity(IncommingInvoices _entity, IncommingInvoiceBO bo, UnitOfWork uow)
        {
            if (_entity == null)
                return ErrorCode.EntityNull;
            if (bo == null)
                return ErrorCode.BoNull;
            if (bo.ProjectId != 0)
                _entity.ProjectId = bo.ProjectId;
            if (bo.ContractID != 0)
                _entity.ContractId = bo.ContractID;
            if (bo.CompanyId != 0)
                _entity.CompanyId = bo.CompanyId;
            _entity.Date = bo.IncommingInvoiceDate;
            _entity.ExternalId = bo.InvoiceExternalId;
            _entity.Price = bo.InvoicePrice;

            var err = HandleRows(_entity, bo.Details);
            if ((err != ErrorCode.Success))
                return err;
            return ErrorCode.Success;
        }

        private static ErrorCode HandleRows(IncommingInvoices _entity, List<IncommingInvoiceDetailBO> rows)
        {
            if ((rows.Count == 0))
                return ErrorCode.Success;
            foreach (var x in rows)
            {
                ErrorCode err;
                if ((x.Id == 0))
                {
                    // insert
                    IncommingInvoiceDetail row = new IncommingInvoiceDetail();
                    err = IncommingInvoiceDetailTranslator.TranslateBOToEntity(row, x);
                    _entity.IncommingInvoiceDetail.Add(row);
                }
                else
                {
                    // update
                    var row = _entity.IncommingInvoiceDetail.FirstOrDefault(f => f.Id == x.Id);
                    if ((row != null))
                        err = IncommingInvoiceDetailTranslator.TranslateBOToEntity(row, x);
                }
            }
            // delete
            List<IncommingInvoiceDetail> delList = new List<IncommingInvoiceDetail>();
            foreach (var x in _entity.IncommingInvoiceDetail)
            {
                if ((!rows.Any(f => f.Id == x.Id)))
                    delList.Add(x);
            }
            foreach (var x in delList)
                _entity.IncommingInvoiceDetail.Remove(x);
            return ErrorCode.Success;
        }
    }
}

File: Service/Translators/IncommingInvoiceTranslator.vb
Imports DAL
Imports BO
Imports System.Text.RegularExpressions
Public Class IncommingInvoiceTranslator
    Friend Shared Function TranslateEntityToBO(_entity As IncommingInvoices, bo As IncommingInvoiceBO) As ErrorCode
        If _entity Is Nothing Then Return ErrorCode.EntityNull
        If bo Is Nothing Then Return ErrorCode.BoNull
        bo.Id = _entity.ID
        bo.ProjectId = _entity.Project.ProjectID
        If Not _entity.Contract Is Nothing Then bo.ContractID = _entity.ContractID
        If Not _entity.CompanyInfo Is Nothing Then bo.CompanyId = _entity.CompanyID
        bo.IncommingInvoiceDate = _entity.Date
        If Not _entity.ExternalID Is Nothing Then bo.InvoiceExternalId = _entity.ExternalID
        bo.InvoicePrice = _entity.Price
        For Each x In _entity.IncommingInvoiceDetail
            Dim bou As New IncommingInvoiceDetailBO
            Dim err = IncommingInvoiceDetailTranslator.TranslateEntityToBO(x, bou)
            bo.Details.Add(bou)
        Next
        Return ErrorCode.Success
    End Function
    Friend Shared Function TranslateBOToEntity(_entity As IncommingInvoices, bo As IncommingInvoiceBO, uow As UnitOfWork) As ErrorCode
        If _entity Is Nothing Then Return ErrorCode.EntityNull
        If bo Is Nothing Then Return ErrorCode.BoNull
        If Not bo.ProjectId = 0 Then _entity.ProjectID = bo.ProjectId
        If Not bo.ContractID = 0 Then _entity.ContractID = bo.ContractID
        If Not bo.CompanyId = 0 Then _entity.CompanyID = bo.CompanyId
        _entity.Date = bo.IncommingInvoiceDate
        _entity.ExternalID = bo.InvoiceExternalId
        _entity.Price = bo.InvoicePrice

        Dim err = HandleRows(_entity, bo.Details)
        If (err <> ErrorCode.Success) Then Return err
        Return ErrorCode.Success
    End Function

    Private Shared Function HandleRows(_entity As IncommingInvoices, rows As List(Of IncommingInvoiceDetailBO)) As ErrorCode
        If (rows.Count = 0) Then Return ErrorCode.Success
        For Each x In rows
            If (x.Id = 0) Then
                'insert
                Dim row As New IncommingInvoiceDetail
                Dim err = IncommingInvoiceDetailTranslator.TranslateBOToEntity(row, x)
                _entity.IncommingInvoiceDetail.Add(row)
            Else
                'update
                Dim row = _entity.IncommingInvoiceDetail.FirstOrDefault(Function(f) f.ID = x.Id)
                If (row IsNot Nothing) Then
                    Dim err = IncommingInvoiceDetailTranslator.TranslateBOToEntity(row, x)
                End If
            End If
        Next
        'delete
        Dim delList As New List(Of IncommingInvoiceDetail)
        For Each x In _entity.IncommingInvoiceDetail
            If (Not rows.Any(Function(f) f.Id = x.ID)) Then
                delList.Add(x)
            End If
        Next
        For Each x In delList
            _entity.IncommingInvoiceDetail.Remove(x)
        Next
        Return ErrorCode.Success
    End Function
End Class
